File: FEMme/src/Mesh.jl
module Mesh
include("NodeArrangement.jl")
#function __reset__(self)
#   """Class resetter. Resets all elements of the class
#   """
#   for i in self.__dict__.keys():
#      self.__dict__[i] = None
#   end
#end

function InferPolynomialDegree(element_type,nodes_per_element)
   """Infer the degree of interpolation (p) based on the shape of
      self.elements

      returns:        [int] polynomial degree
   """
   #@assert element_type!=nothing
   #@assert elements!=nothing

   # check if degree is already asign
   #if degree!=nothing
   #   if isa(degree,Array)
   #      degree = np.asscalar(self.degree)
   #   end
   #   i = degree
   #   if element_type == "tet" && (i+1)*(i+2)*(i+3)/6==size(elements,2)
   #      return degree
   #   end
   #   if element_type == "tri" && (i+1)*(i+2)/2==size(elements,2)
   #      return degree
   #   end
   #end

   p = 0
   if element_type == "tet"
      for i=1:100
         if (i+1)*(i+2)*(i+3)/6==nodes_per_element
            p = i
            break
         end
      end
   elseif element_type == "tri"
      for i=1:100
         if (i+1)*(i+2)/2==nodes_per_element
            p = i
            break
         end
      end
   elseif element_type == "hex"
      for i=1:100
         if (i+1)^3==nodes_per_element
            p = i
            break
         end
      end
   elseif element_type == "quad"
      for i=1:100
         if (i+1)^2==nodes_per_element
            p = i
            break
         end
      end
   elseif element_type == "line"
      for i=1:100
         if i+1==nodes_per_element
            p = i
            break
         end
      end
   elseif element_type == "pent"
      if 5==nodes_per_element
         p = 1
      else
         ErrorException("High order pentagonal elements are not supported yet")
      end
   end
   #degree = p self.
   return p
end

function GetFacesHex(elements)
   """Find all faces (surfaces) in the hexahedral mesh (boundary & interior).
      Sets all_faces property and returns it

      returns:
      arr:         array of all faces
   """
   # DETERMINE DEGREE
   nodes_per_element = size(elements,2)
   p = InferPolynomialDegree("Hex",nodes_per_element)

   # DO NOT COMPUTE IF ALREADY COMPUTED
   #if isa(all_faces,Array)
   #if size(all_faces,1) > 1
      # IF LINEAR VERSION IS COMPUTED, DO COMPUTE HIGHER VERSION
   #   if size(all_faces,2) == 4 && p > 1
   #      pass
   #   else
   #      return all_faces
   #   end
   #end
   #end
   node_arranger = NodeArrangementHex(p)[1]
   fsize = ((p+1)^3)::Int

   # GET ALL FACES FROM THE ELEMENT CONNECTIVITY
   faces = np.concatenate((np.concatenate((
         np.concatenate((np.concatenate((np.concatenate((self.elements[:,node_arranger[0,:]],
         self.elements[:,node_arranger[1,:]]),axis=0),self.elements[:,node_arranger[2,:]]),axis=0),
         self.elements[:,node_arranger[3,:]]),axis=0),self.elements[:,node_arranger[4,:]]),axis=0),
         self.elements[:,node_arranger[5,:]]),axis=0).astype(np.int64)

   # REMOVE DUPLICATES
   all_faces, idx = unique2d(faces,consider_sort=True,order=false,return_index=true)

   face_to_element = zeros(Int,size(all_faces,1),2)
   face_to_element[:,1] =  idx % size(elements,1)
   face_to_element[:,2] =  idx // size(elements,1)

   #self.face_to_element = face_to_element

   return all_faces
end

function GetFaces(element_type)
   @assert element_type!=nothing
   #if element_type == "tet"
   #   self.GetFacesTet()
   if element_type == "hex"
      all_faces = GetFacesHex()
   elseif element_type=="tri" || element_type=="quad"
      ArgumentError("2D mesh does not have faces")
   else
      ArgumentError("Type of element not understood")
   end
   return all_faces
end

function InferNumberOfNodesPerElement(;p=nothing, element_type=nothing)
   """Infers number of nodes per element. If p and element_type are
      not None then returns the number of nodes required for the given
      element type with the given polynomial degree"""

   if p!=nothing && element_type!=nothing
      if element_type=="line"
         return (p+1)::Int
      elseif element_type=="tri"
         return ((p+1)*(p+2)/2)::Int
      elseif element_type=="quad"
         return ((p+1)^2)::Int
      elseif element_type=="tet"
         return ((p+1)*(p+2)*(p+3)/6)::Int
      elseif element_type=="hex"
         return ((p+1)^3)::Int
      else
         ArgumentError("Did not understand element type")
      end
   end
   @assert size(elements,1)!=nothing "Element list is empty" #self.
   return size(elements,2) #self.
end

function ReadGmsh(filename, element_type; read_surface_info=false)
   """Read gmsh (.msh) file"""
   try
      fid = open(filename, "r")
   catch #IOError
      println("File $filename not found.")
      ErrorException("Error at opening the file")
   end

   #if self.elements is not None and self.points is not None:
   #   self.__reset__()
   #end

   bel = -1
   if element_type == "line"
      el = 1
   elseif element_type == "tri"
      el = 2
      bel = 2
   elseif element_type == "quad"
      el = 3
      bel = 3
   elseif element_type == "tet"
      el = 4
      bel = 2
   elseif element_type == "hex"
      el = 5
      bel = 3
   else
      ArgumentError("Element type not understood")
   end

   # NEW FAST READER
   var = 0  # for old gmsh versions - needs checks
   rem_nnode, rem_nelem, rem_faces = convert(Int,1e09), convert(Int,1e09), convert(Int,1e09)
   face_counter = 0
   ndim = 0
   for (line_counter,line) in enumerate(readlines(open(filename)))
      plist = split(line)
      if plist[1] == "Dimension"
         ndim = parse(Int,plist[2]) #self.
      elseif plist[1] == "Vertices"
         rem_nnode = line_counter+1
         continue
      elseif plist[1] == "\$Nodes"
         rem_nnode = line_counter+1
         continue
      elseif plist[1] == "Triangles"
         rem_faces = line_counter+1
         continue
      elseif plist[1] == "Tetrahedra"
         rem_nelem = line_counter+1
         continue
      elseif plist[1] == "\$Elements"
         rem_nelem = line_counter+1
         var = 1
         continue
      end
      if rem_nnode == line_counter
         nnode = parse(Int,plist[1]) #self.
      end
      if rem_nnode+1 == line_counter && ndim == 0
         ndim = size(plist,1)-1
      end
      if rem_faces == line_counter
         face_counter = parse(Int,plist[1])
      end
      if rem_nelem == line_counter
         nelem = parse(Int,plist[1]) #self.
         break
      end
   end

   ne = InferNumberOfNodesPerElement(p=1,element_type=element_type)
   # Re-read
   points, elements, faces, face_to_surface = Float64[],Int[], Int[], Int[]
   elem_counter = 0 #to correct the number of elements
   face_counter = 0
   for (line_counter,line) in enumerate(readlines(open(filename)))
      global nf
      plist = split(line)
      if var==0
         if line_counter>rem_nnode && line_counter<nnode+rem_nnode+1
            append!(points,[parse(Float64,i) for i in plist[1:3]])
         end
         if line_counter>rem_nelem && line_counter<nelem+rem_nelem+1
            append!(elements,[parse(Int,i) for i in plist[1:4]])
         end
      elseif var==1
         if line_counter>rem_nnode && line_counter<nnode+rem_nnode+1
            append!(points,[parse(Float64,i) for i in plist[2:end]])
         end
         if line_counter>rem_nelem && line_counter<nelem+rem_nelem+1
            if parse(Int,plist[2]) == el
               elem_counter += 1
               append!(elements,[parse(Int,i) for i in plist[end-ne+1:end]])
            end
            # WRITE SURFACE INFO - CERTAINLY ONLY IF ELEMENT TYPE IS QUADS/TRIS
            if read_surface_info
               if parse(Int,plist[2]) == bel
                  nf = size(plist,1)-5
                  face_counter += 1
                  append!(faces,[parse(Int,i) for i in plist[6:end]])
                  append!(face_to_surface,parse(Int,plist[5]))
               end
            end
         end
      end
   end

   points = reshape(points,ndim,nnode) #self.
   points = transpose(points)
   points = copy(points)
   elements = reshape(elements,ne,elem_counter) #self.
   elements = transpose(elements)
   elements = copy(elements)
   # CORRECT
   nelem = size(elements,1) #self.
   nnode = size(points,1) #self.
   if nelem == 0 #self.
      ArgumentError("msh file does not contain $element_type elements")
   end
   if read_surface_info
      faces = reshape(faces,nf,face_counter) #self.
      faces = transpose(faces)
      faces = copy(faces)
      #self.face_to_surface = face_to_surface
      # CHECK IF FILLED
      if size(face_to_surface,1)==0
         face_to_surface = nothing
      end
   end
   # Reduce dimension if Z coordinates are zero
   if size(points,2) == 3 #self.
      if isapprox(points[:,3],zeros(Float64,size(points,1)))
         points = points[:,1:2]
      end
   end
   # Get boundary data of the mesh
   #if element_type == "tri" || element_type == "quad"
   #   self.GetEdges()
   #   self.GetBoundaryEdges()
   if element_type == "tet" || element_type == "hex"
      all_faces = GetFaces(element_type)
      GetBoundaryFaces()
      GetBoundaryEdges()
   end
   return
end

function Read(;filename=nothing, element_type="tri", reader_type=nothing, reader_type_format=nothing,
   reader_type_version=nothing, order=0, read_surface_info=false)
   """Convenience mesh reader method to dispatch call to subsequent apporpriate methods"""

   if !isa(filename,String)
      ArgumentError("filename must be a string")
      return
   end
   if reader_type != nothing
      if !isa(filename,String)
         ArgumentError("filename must be a string")
         return
      end
   end

   if reader_type==nothing
      if endswith(filename,"msh")
         reader_type = "gmsh"
      elseif endswith(filename,"obj")
         reader_type = "obj"
      elseif endswith(filename,"unv")
         reader_type = "unv"
      elseif endswith(filename,"fro")
         reader_type = "fro"
      elseif endswith(filename,"dat")
         ErrorException("A reader for this mesh is still not implemented")
         #for key in kwargs.keys()
         #   inkey = insensitive(key)
         #   if "connectivity" in inkey and "delimiter" not in inkey
         #      reader_type = "read_separate"
         #      break
         #   end
         #end
      else
         ArgumentError("Mesh file format was not understood. Please specify it using reader_type keyword")
      end
   end

   if reader_type=="salome"
      ErrorException("A reader for this mesh is still not implemented")
      #ReadSalome(filename, element_type=element_type, read_surface_info=read_surface_info)
   elseif reader_type=="GID"
      ErrorException("A reader for this mesh is still not implemented")
      #ReadGIDMesh(filename, element_type, order)
   elseif reader_type=="gmsh"
      ReadGmsh(filename, element_type, read_surface_info=read_surface_info)
   elseif reader_type=="obj"
      ErrorException("A reader for this mesh is still not implemented")
      #ReadOBJ(filename, element_type=element_type, read_surface_info=read_surface_info)
   elseif reader_type=="fenics"
      ErrorException("A reader for this mesh is still not implemented")
      #ReadFenics(filename, element_type)
   elseif reader_type=="vtu"
      ErrorException("A reader for this mesh is still not implemented")
      #ReadVTK(filename)
   elseif reader_type=="unv"
      ErrorException("A reader for this mesh is still not implemented")
      #ReadUNV(filename, element_type)
   elseif reader_type=="fro"
      ErrorException("A reader for this mesh is still not implemented")
      #ReadFRO(filename, element_type)
   elseif reader_type=="read_separate"
      ErrorException("A reader for this mesh is still not implemented")
      # READ MESH FROM SEPARATE FILES FOR CONNECTIVITY AND COORDINATES
      #from Florence.Utils import insensitive
      # return insensitive(kwargs.keys())
      #for key in kwargs.keys()
      #   inkey = insensitive(key)
      #   if "connectivity" in inkey and "delimiter" not in inkey
      #      connectivity_file = kwargs.get(key)
      #   end
      #   if "coordinate" in insensitive(key) and "delimiter" not in inkey
      #      coordinates_file = kwargs.get(key)
      #   end
      #end
      #self.ReadSeparate(connectivity_file,coordinates_file,element_type,
      #   delimiter_connectivity=",",delimiter_coordinates=",")
   elseif reader_type=="ReadHDF5"
      ErrorException("A reader for this mesh is still not implemented")
      #ReadHDF5(filename)
   end

   #self.nnode = self.points.shape[0]
   # MAKE SURE MESH DATA IS CONTIGUOUS
   #self.points   = np.ascontiguousarray(self.points)
   #self.elements = np.ascontiguousarray(self.elements)
   return
end #function

end #module

File: FEMpy/Florence/MeshGeneration/Mesh.py
#from __future__ import print_function, division
import os, sys, warnings, platform
#from time import time
import numpy as np
#if "PyPy" not in platform.python_implementation():
#    from scipy.io import loadmat, savemat
#from Florence.Tensor import makezero, itemfreq, unique2d, in2d
#from Florence.Utils import insensitive
#from .vtk_writer import write_vtu
#try:
#    import meshpy.triangle as triangle
#    has_meshpy = True
#except ImportError:
#    has_meshpy = False
#from .HigherOrderMeshing import *
from .NodeArrangement import *
#from .GeometricPath import *
#from warnings import warn
#from copy import deepcopy



"""
Mesh class providing most of the pre-processing functionalities of the Core module

Roman Poya - 13/06/2015
"""



class Mesh(object):

    """Mesh class provides the following functionalities:
    1. Generating higher order meshes based on a linear mesh, for tris, tets, quads and hexes
    2. Generating linear tri and tet meshes based on meshpy back-end
    3. Generating linear tri meshes based on distmesh back-end
    4. Finding bounary edges and faces for tris and tets, in case they are not provided by the mesh generator
    5. Reading Salome meshes in binary (.dat/.txt/etc) format
    6. Reading gmsh files .msh
    7. Checking for node numbering order of elements and fixing it if desired
    8. Writing meshes to unstructured vtk file format (.vtu) in xml and binary formats,
        including high order elements
    """

    def __init__(self, element_type=None):
        super(Mesh, self).__init__()
        # self.faces and self.edges ARE BOUNDARY FACES
        # AND BOUNDARY EDGES, RESPECTIVELY

        self.degree = None
        self.ndim = None
        self.edim = None
        self.nelem = None
        self.nnode = None

        self.elements = None
        self.points = None
        self.corners = None
        self.edges = None
        self.faces = None
        self.element_type = element_type

        self.face_to_element = None
        self.edge_to_element = None
        self.boundary_edge_to_element = None
        self.boundary_face_to_element = None
        self.all_faces = None
        self.all_edges = None
        self.interior_faces = None
        self.interior_edges = None
        # TYPE OF BOUNDARY FACES/EDGES
        self.boundary_element_type = None

        # FOR GEOMETRICAL CURVES/SURFACES
        self.edge_to_curve = None
        self.face_to_surface = None


        self.spatial_dimension = None
        self.reader_type = None
        self.reader_type_format = None
        self.reader_type_version = None
        self.writer_type = None

        self.filename = None
        # self.has_meshpy = has_meshpy

    def GetFaces(self):
        assert self.element_type is not None
        if self.element_type == "tet":
            self.GetFacesTet()
        elif self.element_type == "hex":
            self.GetFacesHex()
        elif self.element_type=="tri" or self.element_type=="quad":
            raise ValueError("2D mesh does not have faces")
        else:
            raise ValueError('Type of element not understood')
        return self.all_faces

    def GetFacesHex(self):
        """Find all faces (surfaces) in the hexahedral mesh (boundary & interior).
            Sets all_faces property and returns it

        returns:

            arr:            numpy ndarray of all faces

        """

        # DETERMINE DEGREE
        p = self.InferPolynomialDegree()

        # DO NOT COMPUTE IF ALREADY COMPUTED
        if isinstance(self.all_faces,np.ndarray):
            if self.all_faces.shape[0] > 1:
                # IF LINEAR VERSION IS COMPUTED, DO COMPUTE HIGHER VERSION
                if self.all_faces.shape[1] == 4 and p > 1:
                    pass
                else:
                    return self.all_faces

        node_arranger = NodeArrangementHex(p-1)[0]
        fsize = int((p+1)**3)

        # GET ALL FACES FROM THE ELEMENT CONNECTIVITY
        faces = np.concatenate((np.concatenate((
                np.concatenate((np.concatenate((np.concatenate((self.elements[:,node_arranger[0,:]],
                self.elements[:,node_arranger[1,:]]),axis=0),self.elements[:,node_arranger[2,:]]),axis=0),
                self.elements[:,node_arranger[3,:]]),axis=0),self.elements[:,node_arranger[4,:]]),axis=0),
                self.elements[:,node_arranger[5,:]]),axis=0).astype(np.int64)

        # REMOVE DUPLICATES
        self.all_faces, idx = unique2d(faces,consider_sort=True,order=False,return_index=True)

        face_to_element = np.zeros((self.all_faces.shape[0],2),np.int64)
        face_to_element[:,0] =  idx % self.elements.shape[0]
        face_to_element[:,1] =  idx // self.elements.shape[0]

        self.face_to_element = face_to_element

        return self.all_faces

    def Read(self, filename=None, element_type="tri", reader_type=None, reader_type_format=None,
        reader_type_version=None, order=0, read_surface_info=False, **kwargs):
        """Convenience mesh reader method to dispatch call to subsequent apporpriate methods"""

        if not isinstance(filename,str):
            raise ValueError("filename must be a string")
            return
        if reader_type is not None:
            if not isinstance(filename,str):
                raise ValueError("filename must be a string")
                return

        if reader_type is None:
            if filename.split('.')[-1] == "msh":
                reader_type = "gmsh"
            elif filename.split('.')[-1] == "obj":
                reader_type = "obj"
            elif filename.split('.')[-1] == "unv":
                reader_type = "unv"
            elif filename.split('.')[-1] == "fro":
                reader_type = "fro"
            elif filename.split('.')[-1] == "dat":
                for key in kwargs.keys():
                    inkey = insensitive(key)
                    if "connectivity" in inkey and "delimiter" not in inkey:
                        reader_type = "read_separate"
                        break
            if reader_type is None:
                raise ValueError("Mesh file format was not undertood. Please specify it using reader_type keyword")


        self.filename = filename
        self.reader_type = reader_type
        self.reader_type_format = reader_type_format
        self.reader_type_version = reader_type_version

        if self.reader_type is 'salome':
            self.ReadSalome(filename, element_type=element_type, read_surface_info=read_surface_info)
        elif reader_type is 'GID':
            self.ReadGIDMesh(filename, element_type, order)
        elif self.reader_type is 'gmsh':
            self.ReadGmsh(filename, element_type=element_type, read_surface_info=read_surface_info)
        elif self.reader_type is 'obj':
            self.ReadOBJ(filename, element_type=element_type, read_surface_info=read_surface_info)
        elif self.reader_type is 'fenics':
            self.ReadFenics(filename, element_type)
        elif self.reader_type is 'vtu':
            self.ReadVTK(filename)
        elif self.reader_type is 'unv':
            self.ReadUNV(filename, element_type)
        elif self.reader_type is 'fro':
            self.ReadFRO(filename, element_type)
        elif self.reader_type is 'read_separate':
            # READ MESH FROM SEPARATE FILES FOR CONNECTIVITY AND COORDINATES
            from Florence.Utils import insensitive
            # return insensitive(kwargs.keys())
            for key in kwargs.keys():
                inkey = insensitive(key)
                if "connectivity" in inkey and "delimiter" not in inkey:
                    connectivity_file = kwargs.get(key)
                if "coordinate" in insensitive(key) and "delimiter" not in inkey:
                    coordinates_file = kwargs.get(key)

            self.ReadSeparate(connectivity_file,coordinates_file,element_type,
                delimiter_connectivity=',',delimiter_coordinates=',')
        elif self.reader_type is 'ReadHDF5':
            self.ReadHDF5(filename)

        self.nnode = self.points.shape[0]
        # MAKE SURE MESH DATA IS CONTIGUOUS
        self.points   = np.ascontiguousarray(self.points)
        self.elements = np.ascontiguousarray(self.elements)
        return

    def ReadGmsh(self, filename, element_type, read_surface_info=False):
        """Read gmsh (.msh) file"""

        try:
            fid = open(filename, "r")
        except IOError:
            print("File '%s' not found." % (filename))
            sys.exit()

        msh_version = None
        # CHECK MSH FILE VERSION
        if "MeshFormat" in fid.readline():
            msh_version = int(np.floor(float(fid.readline().split(" ")[0])))
            if 4 != msh_version and 2 != msh_version:
                raise IOError("Only ASCII version 2 and 4 (>=4.1) .msh file formats are supported")
        if 4 != msh_version and 2 != msh_version:
            raise IOError("Only ASCII version 2 and 4 (>=4.1) .msh file formats are supported")
        fid.close()

        if self.elements is not None and self.points is not None:
            self.__reset__()

        self.filename = filename

        bel = -1
        if element_type == "line":
            el = 1
        elif element_type == "tri":
            el = 2
            bel = 2
        elif element_type == "quad":
            el = 3
            bel = 3
        elif element_type == "tet":
            el = 4
            bel = 2
        elif element_type == "hex":
            el = 5
            bel = 3
        else:
            raise ValueError("Element type not understood")


        # NEW FAST READER
        var = 0 # for old gmsh versions - needs checks
        node_blocks, elem_blocks, face_blocks = None, None, None
        rem_nnode, rem_nelem, rem_faces = int(1e09), int(1e09), int(1e09)
        face_counter = 0
        for line_counter, line in enumerate(open(filename)):
            item = line.rstrip()
            plist = item.split()
            if plist[0] == "Dimension":
                self.ndim = plist[1]
            elif plist[0] == "Vertices":
                rem_nnode = line_counter+1
                continue
            elif plist[0] == "$Nodes":
                rem_nnode = line_counter+1
                continue
            elif plist[0] == "Triangles":
                rem_faces = line_counter+1
                continue
            elif plist[0] == "Tetrahedra":
                rem_nelem = line_counter+1
                continue
            elif plist[0] == "$Elements":
                rem_nelem = line_counter+1
                var = 1
                continue

            if msh_version == 2:
                if rem_nnode == line_counter:
                    self.nnode = int(plist[0])
                if rem_faces == line_counter:
                    face_counter = int(plist[0])
                if rem_nelem == line_counter:
                    self.nelem = int(plist[0])
                    break
            else:
                if rem_nnode == line_counter:
                    node_blocks, self.nnode = int(plist[0]), int(plist[1])
                if rem_faces == line_counter:
                    face_blocks, face_counter = int(plist[0]), int(plist[1])
                if rem_nelem == line_counter:
                    elem_blocks, self.nelem = int(plist[0]), int(plist[1])
                    break

        points, elements, faces, face_to_surface = [],[], [], []
        if msh_version == 2:
            # RE-READ
            ns = self.InferNumberOfNodesPerElement(p=1,element_type=element_type)
            for line_counter, line in enumerate(open(filename)):
                item = line.rstrip()
                plist = item.split()
                if var == 0:
                    if line_counter > rem_nnode and line_counter < self.nnode+rem_nnode+1:
                        points.append([float(i) for i in plist[:3]])
                    if line_counter > rem_nelem and line_counter < self.nelem+rem_nelem+1:
                        elements.append([int(i) for i in plist[:4]])
                elif var == 1:
                    if line_counter > rem_nnode and line_counter < self.nnode+rem_nnode+1:
                        points.append([float(i) for i in plist[1:]])
                    if line_counter > rem_nelem and line_counter < self.nelem+rem_nelem+1:
                        if int(plist[1]) == el:
                            elements.append([int(i) for i in plist[-ns:]])

                        # READ SURFACE INFO - CERTAINLY ONLY IF SURFACE ELEMENT TYPE IS QUADS/TRIS
                        if read_surface_info:
                            if int(plist[1]) == bel:
                                faces.append([int(i) for i in plist[5:]])
                                face_to_surface.append(int(plist[4]))


        elif msh_version == 4:
            # RE-READ
            fid = open(filename)
            content = fid.readlines()

            # READ NODES
            nodes_content = content[rem_nnode+1:2*self.nnode+node_blocks+rem_nnode+1]
            incrementer, line_number = 0, 0
            # LOOP OVER BLOCKS
            for i in range(node_blocks):
                incrementer = int(nodes_content[line_number].rstrip().split()[3])
                # LOOP OVER NODES OF EACH BLOCK
                for j in range(line_number+1, line_number+2*incrementer+1):
                    plist = nodes_content[j].rstrip().split()
                    if len(plist) == 1:
                        continue
                    points.append([float(plist[k]) for k in range(0,len(plist))])
                line_number += 2*incrementer + 1

            # READ ELEMENTS
            elems_content = content[rem_nelem+1:self.nelem+elem_blocks+rem_nelem+1]
            incrementer, line_number = 0, 0
            # LOOP OVER BLOCKS
            for i in range(elem_blocks):
                incrementer = int(elems_content[line_number].rstrip().split()[3])
                if el == int(elems_content[line_number].rstrip().split()[2]):
                    # LOOP OVER ELEMENTS OF EACH BLOCK
                    for j in range(line_number+1, line_number+incrementer+1):
                        plist = elems_content[j].rstrip().split()
                        elements.append([int(plist[k]) for k in range(1,len(plist))])
                line_number += incrementer + 1

            if read_surface_info:
                # READ FACES
                incrementer, line_number = 0, 0
                # LOOP OVER BLOCKS
                for i in range(elem_blocks):
                    incrementer = int(elems_content[line_number].rstrip().split()[3])
                    surface_tag = int(elems_content[line_number].rstrip().split()[1])
                    if bel == int(elems_content[line_number].rstrip().split()[2]):
                        # LOOP OVER FACES OF EACH BLOCK
                        for j in range(line_number+1, line_number+incrementer+1):
                            plist = elems_content[j].rstrip().split()
                            faces.append([int(plist[k]) for k in range(1,len(plist))])
                            face_to_surface.append(surface_tag)
                    line_number += incrementer + 1


        self.points = np.array(points,copy=True)
        self.elements = np.array(elements,copy=True) - 1
        # CORRECT
        self.nelem = self.elements.shape[0]
        self.nnode = self.points.shape[0]
        if self.nelem == 0:
            raise ValueError("msh file does not contain {} elements".format(element_type))

        if read_surface_info:
            self.faces = np.array(faces,copy=True) - 1
            self.face_to_surface = np.array(face_to_surface, dtype=np.int64, copy=True).flatten()
            self.face_to_surface -= 1
            # CHECK IF FILLED
            if isinstance(self.face_to_surface,list):
                if not self.face_to_surface:
                    self.face_to_surface = None
            elif isinstance(self.face_to_surface,np.ndarray):
                if self.face_to_surface.shape[0]==0:
                    self.face_to_surface = None

        if self.points.shape[1] == 3:
            if np.allclose(self.points[:,2],0.):
                self.points = np.ascontiguousarray(self.points[:,:2])

        self.element_type = element_type
        if self.element_type == "tri" or self.element_type == "quad":
            self.GetEdges()
            self.GetBoundaryEdges()
        elif self.element_type == "tet" or self.element_type == "hex":
            self.GetFaces()
            self.GetBoundaryFaces()
            self.GetBoundaryEdges()

        return

    def InferPolynomialDegree(self):
        """Infer the degree of interpolation (p) based on the shape of
            self.elements

            returns:        [int] polynomial degree
            """

        assert self.element_type is not None
        assert self.elements is not None

        if self.degree is not None:
            if isinstance(self.degree,np.ndarray):
                self.degree = np.asscalar(self.degree)
            i = self.degree
            if self.element_type == "tet" and (i+1)*(i+2)*(i+3)/6==self.elements.shape[1]:
                return self.degree
            if self.element_type == "tri" and (i+1)*(i+2)/2==self.elements.shape[1]:
                return self.degree


        p = 0
        if self.element_type == "tet":
            for i in range(100):
                if (i+1)*(i+2)*(i+3)/6==self.elements.shape[1]:
                    p = i
                    break

        elif self.element_type == "tri":
            for i in range(100):
                if (i+1)*(i+2)/2==self.elements.shape[1]:
                    p = i
                    break

        elif self.element_type == "hex":
            for i in range(100):
                if int((i+1)**3)==self.elements.shape[1]:
                    p = i
                    break

        elif self.element_type == "quad":
            for i in range(100):
                if int((i+1)**2)==self.elements.shape[1]:
                    p = i
                    break

        elif self.element_type == "line":
            for i in range(100):
                if int(i+1)==self.elements.shape[1]:
                    p = i
                    break

        elif self.element_type == "pent":
            if 5==self.elements.shape[1]:
                p = 1
            else:
                raise NotImplementedError("High order pentagonal elements are not supported yet")

        self.degree = p
        return p

    def InferNumberOfNodesPerElement(self, p=None, element_type=None):
        """Infers number of nodes per element. If p and element_type are
            not None then returns the number of nodes required for the given
            element type with the given polynomial degree"""

        if p is not None and element_type is not None:
            if element_type=="line":
                return int(p+1)
            elif element_type=="tri":
                return int((p+1)*(p+2)/2)
            elif element_type=="quad":
                return int((p+1)**2)
            elif element_type=="tet":
                return int((p+1)*(p+2)*(p+3)/6)
            elif element_type=="hex":
                return int((p+1)**3)
            else:
                raise ValueError("Did not understand element type")

        assert self.elements.shape[0] is not None
        return self.elements.shape[1]

